File: javascript/secret-handshake/secret-handshake.js
const OPERATIONS = {
    1: (ops) => ops.push('wink'),
    2: (ops) => ops.push('double blink'),
    4: (ops) => ops.push('close your eyes'),
    8: (ops) => ops.push('jump'),
    16: (ops) => ops.reverse()
};

export function secretHandshake(input) {
    const num = Number(input);
    if (isNaN(num)) {
        throw new Error('Handshake must be a number');
    }

    const ops = [];
    for (const opBit of Object.keys(OPERATIONS)) {
        if ((num & Number(opBit)) != 0) {
            OPERATIONS[opBit](ops);
        }
    }
    return ops;
}

File: haskell/secret-handshake/_iterations/2/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits

handshake :: Int -> [String]
handshake n = reverseAction n . handshake' n actions $ []
  where
    actions :: [String]
    actions = ["wink", "double blink", "close your eyes", "jump"]

    handshake' :: Int -> [String] -> [String] -> [String]
    handshake' _ [] out = out
    handshake' 0 _ out = out
    handshake' n (a:as) out
             | testBit n 0 = handshake' n' as (a:out)
             | otherwise = handshake' n' as out
                 where n' = shiftR n 1

    reverseAction :: Int -> [String] -> [String]
    reverseAction n out
                | testBit n 4 = out
                | otherwise = reverse out
